feat: Unify low-level operation with consistent HAL interface

splasky opened the pull request:
Add hardware abstraction layer for create unify devices and
device operators. See `HAL.md` for how to implement device.

Closes #63

jserv commented:
You have to create corresponding issues which describe the action items and executions plans. Then, create new pull request(s) such as this one to close the issues.

jserv commented on Makefile:
```diff
@@ -25,25 +26,21 @@ CFLAGS = -Wall -fPIC -DHAVE_CONFIG_H -D_U_="__attribute__((unused))" -g3 -DDEBUG
 
 .SUFFIXES:.c .o
-.PHONY: all clean test pre-build
+.PHONY: all clean test pre-build mbedtls_make hal
```
Why `mbedtls_make`?

jserv commented:
You shall provide documentation for HAL.

jserv commented:
We are engineers instead of code monkeys, OK?

jserv commented:
Use uppercase form "HAL" for each subject which refers to HAL.

jserv commented:
Rebasing is required.

jserv commented:
FYI. AOSP has a neat example for HAL design: [liblog/logd_reader.cpp](https://android.googlesource.com/platform/system/core/+/refs/tags/android-8.1.0_r74/liblog/logd_reader.c).
Check [liblog](https://android.googlesource.com/platform/system/core/+/refs/tags/android-8.1.0_r74/liblog) as well.

howjmay commented on hal/device.c:
```diff
@@ -0,0 +1,50 @@
+  if (dv->next) return -DEVICE_ERROR;
+  p = find_device(dv->name, strlen(dv->name));
+  if (*p)
+    res = -DEVICE_ERROR;
```
use curl braces here

splasky replied:
What is curl braces?

howjmay replied:
this `{`

howjmay commented on hal/device.c:
```diff
@@ -0,0 +1,50 @@
+int register_device(struct device_type *dv) {
+  int res = 0;
+  struct device_type **p;
+  if (dv->next) return -DEVICE_ERROR;
```
use curl braces here

howjmay commented on hal/device.c:
```diff
@@ -0,0 +1,50 @@
+  if (*p)
+    res = -DEVICE_ERROR;
+  else
+    *p = dv;
```
use curl braces here

howjmay commented on hal/device.h:
```diff
@@ -0,0 +1,62 @@
+enum { DEVICE_OK, DEVICE_ERROR };
+
+struct device_operations {
+  int (*init)(void);            /**< init for device */
```
Return error code should be in `retcode_t`

howjmay commented on hal/device.h:
```diff
@@ -0,0 +1,62 @@
+struct device_operations {
+  int (*init)(void);            /**< init for device */
+  void (*fini)(void);           /**< destructor for device  */
+  int (*get_key)(uint8_t *);    /**< get device private key */
```
ditto

howjmay commented on hal/device.h:
```diff
@@ -0,0 +1,62 @@
+};
+
+struct uart_operations {
+  int (*init)(const char *device);
```
ditto

splasky replied:
This return is the file descriptor.

howjmay commented on hal/device.h:
```diff
@@ -0,0 +1,62 @@
+};
+
+device_t *get_device(const char *);
+int register_device(struct device_type *dv);
```
ditto

jserv commented on utils/defined_error.h:
```diff
@@ -23,6 +23,14 @@ typedef enum {
+  DEVICE_REG,     /**< Register device error */
+  DEVICE_UNREG,   /**< Unregister device error */
+  UART_INIT,      /**< Initialize uart error */
+  LE_NO_MEMORY,   /**< No memory error */
```
Shall we expose enumerations with `LE_` prefix?

jserv replied:
We have to decouple Legato with HAL interface.

jserv commented on hal/device.c:
```diff
@@ -0,0 +1,52 @@
+retcode_t unregister_device(struct device_type *dv) {
+  struct device_type **tmp;
+  tmp = &devices;
+  while (*tmp) {
```
Rewrite with `for` statement, which can be shorter and more elegant.

jserv commented on hal/device.c:
```diff
@@ -0,0 +1,52 @@
+  return p;
+}
+
+device_t *get_device(const char *type) {
```
It is confusing to have `get_device` and `get_device_id`. You have to distinguish.

jserv commented:
Fix conflicts.

jserv reviewed (changes requested):
Actions:
1. Mention the issue number that you wish to close within pull request descriptions;
2. Add HAL design documentation in top-level directory;

jserv commented on hal/device.c:
```diff
@@ -0,0 +1,49 @@
+#include "device.h"
```
Add copyright notice.

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+ * - NULL on failed
+ * @see #device_type
+ */
+device_t *device(const char *device);
```
The naming is misleading.

jserv commented on hal/device.c:
```diff
@@ -0,0 +1,49 @@
+static struct device_type **find_device(const char *name, unsigned len) {
+  struct device_type **p;
+  for (p = &devices; *p; p = &(*p)->next)
+    if (strlen((*p)->name) == len && strncmp((*p)->name, name, len) == 0) break;
```
Change from
`strncmp((*p)->name, name, len) == 0`
to
`!strncmp((*p)->name, name, len)`

jserv commented on hal/device.c:
```diff
@@ -0,0 +1,49 @@
+  }
+  p = find_device(type, strlen(type));
+  if (*p) {
+    fprintf(stderr, "Device type %s not found", type);
```
Shall we integrate logger here?

splasky replied:
I think we should focus on HAL subject here. Logger and and other logging message we will resolve by #34 .

jserv replied:
Add `TODO` and/or `FIXME`.

jserv commented on hal.md:
```diff
@@ -0,0 +1,117 @@
+How to porting TA-endpoint on your device
+===
+###### tags: `TA-endpoint`
+Written by HY Chang\<hychang@biilabs.io\>
```
Don't mention yourself. It is a comprehensive work.

jserv commented on hal.md:
```diff
@@ -0,0 +1,117 @@
+How to porting TA-endpoint on your device
+===
+###### tags: `TA-endpoint`
```
Don't use `tags`.

jserv commented on hal.md:
```diff
@@ -0,0 +1,117 @@
+How to porting TA-endpoint on your device
```
The subject should be covering HAL.

jserv replied:
You need to address motivation, concepts, feasible terms, etc.

splasky replied:
Should we change the hal.md to HAL.md?

jserv replied:
go ahead.

jserv replied:
Use fully-qualified subject as well.

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+   * - #UNAVAILABLE on secure storage is currently unavailable
+   * - #FAULT on some other error
+   */
+  retcode_t (*write)(const char *name, const uint8_t *bufPtr, size_t bufSize);
```
No camelCase.

jserv commented on hal.md:
````diff
@@ -0,0 +1,117 @@
+`$(CC)`,`$(CFLAGS)`,`$(INCLUDES)` is provide by TA-endpoint. Don't define same flags inside your Makefile.
+
+A example for `impl.c`:
+```c=
````
Use strict `c` for syntax highlighting.

jserv commented on hal.md:
````diff
@@ -0,0 +1,117 @@
+mydevice.o: impl.c
+        $(CC) $(CFLAGS) $(INCLUDES) -c $^ -o $@
+```
+`$(CC)`,`$(CFLAGS)`,`$(INCLUDES)` is provide by TA-endpoint. Don't define same flags inside your Makefile.
````
Fix grammar.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,164 @@
+TA-endpoint and HAL overview
```
Use starting `#`

jserv commented on HAL.md:
```diff
@@ -0,0 +1,164 @@
+===
+
+# Abstract 
+The TA-endpoint uses HAL(hardware abstract layer) to encapsulate the logic of driver of kinds of device. HAL layer support such as uart, GPIO, secure storage and other low-level operators used in TA-endpoint. The TA-endpoint is a daemon which base on Linux which offer monitor for device and can send the message to tangle-accelerator. HAL layer lets user only needs to implement their device operat ...
```
Change the first line to
> Hardware Abstraction Layer (HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations.

Don't mention TA-endpoint anymore since we are going to consolidate.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,164 @@
+* third_party: Third party directory.
+* utils: Utility functions for TA-endpoint.
+
+# Create folder for your device
```
Never say "folder."
It is indeed a directory.

jserv replied:
Avoid referring to "you" or "your."

jserv commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract 
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer support such as uart, GPIO, secure storage and other low-level operators used in TA-endpoint. The TA-endpoint is a daemon which base on Linux which offer monitor for device and can send the message to tangle-acce ...
```
Change `TA-endpoint` to `the endpoint`.

jserv replied:
Fix grammar.

marktwtn commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+Here are some functions needed to use inside `impl.c`:
+* register_device : register device on startup
+* unregister_device : unregister device
+* DECLARE_DEVICE : this must been declare inside `impl.c` 
```
this must be declared inside `impl.c`.

marktwtn commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+
+all: $(DEVICES_OBJ)
+... 
+mydevice.o: device.o
+    $(MAKE) -C ../devices/mydevice
```
Should the directory naming needs to be unified with the `line 18` in `HAL.mk` to avoid confusion?

splasky replied:
I think unify them would better.

marktwtn commented on HAL.md:
````diff
@@ -0,0 +1,117 @@
+```
+
+### Implement new device
+Implement new device which create under devices directory, edit Makefile, example device name as mydevice:
````
Implement new device which `is created` under the devices directory `and edit the` Makefile.
`The example device is named` as mydevice:

marktwtn commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+## Abstract 
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer support such as uart, GPIO, secure storage and other low-level operators used in TA-endpoint. The TA-endpoint is a daemon which base on Linux which offer monitor for device and can send the message to tangle-acce ...
+
+## Directory Overview
```
`overview` for unification.

marktwtn commented on HAL.md:
````diff
@@ -0,0 +1,117 @@
+```
+mkdir devices/<my_device>
+```
+* Create `impl.c` and `Makefile` under new device directory.
+* Include `device.h` headers into new device header file.
````
Should we mention the header file `impl.h`?

splasky replied:
`impl.h` is not necessary. If users need it, they can create it by themselves.

marktwtn replied:
But the `impl.h` is mentioned at the line `78` of the example code of `impl.c` in the `HAL.md`.

marktwtn commented on hal/Makefile:
```diff
@@ -0,0 +1,12 @@
+	$(CC) $(CFLAGS) $(INCLUDES) -c $^
+
+clean:
+	find $(DEVICES_PATH) -name "*.o" -exec rm -f {} \;
```
Where is the **DEVICES_PATH** variable?

marktwtn commented on utils/defined_error.h:
```diff
@@ -24,6 +24,14 @@ typedef enum {
+  DEVICE_REG,     /**< register device error */
+  DEVICE_UNREG,   /**< unregister device error */
+  UART_INIT,    /**< initialize uart error */
+  NO_MEMORY,    /**< no enough memory error */
```
I think the comment should be `not enough`.

marktwtn commented on utils/defined_error.h:
```diff
@@ -24,6 +24,14 @@ typedef enum {
   RET_HTTP_CERT,    /**< failed on x509 cert parse */
   RET_HTTP_CONNECT, /**< failed on HTTP initial connection */
   RET_HTTP_SSL,     /**< failed on setting ssl config */
+  DEVICE_REG,     /**< register device error */
+  DEVICE_UNREG,   /**< unregister device error */
+  UART_INIT,    /**< initialize uart error */
+  NO_MEMORY,    /**< no enough memory error */
+  OVERFLOW,     /**< overflow error */
+  NOT_FOUND,    /**< item not found error */
+  UNAVAILABLE,  /**< item not available */
+  FAULT         /** some other fault */
```
Should the naming starts with **RET_** prefix?

marktwtn commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+   * @return
+   * - #RET_OK on success
+   * - #NO_MEMORY on no memory error
+   * - #UNAVAILABLE on secure storage is currently unavailable
```
I think the grammar is a little weird.
Change to `unavailable secure storage` for simplicity?

marktwtn commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+   * - #OVERFLOW error on the buffer is too small
+   * - #NOT_FOUND error on the item not found inside secure storage
+   * - #UNAVAILABLE error if the storage is currently unavailable
+   * - #FAULT if there was some other error
```
`is`

marktwtn replied:
I think using `are` might be better.

marktwtn commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+   * - #RET_OK on success
+   * - #NOT_FOUND error on the item not found inside secure storage
+   * - #UNAVAILABLE error if the storage is currently unavailable
+   * - #FAULT if there was some other error
```
Ditto.

marktwtn commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+  const struct device_operations *op;            /**< device operations handler */
+  const struct uart_operations *uart;            /**< uart operations handler */
+  const struct secure_store_operations *sec_ops; /**< secure storage operations handler */
+  device_t *next;                                /**< Pointer to next device type, don't use this directly */
```
`Pointer` -> `pointer`.

marktwtn replied:
The data type `device_t` is defined at?

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+
+## Quick view inside device.h
+### Abstraction 
+Here are some operations needed to implement inside new device:
```
Here are some operations needed to be implemented for new device:

howjmay commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+typedef struct device_type device_t;
+
+struct device_operations {
+  retcode_t (*init)(void);            /**< initialize for device */
```
initialize device

howjmay commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+
+struct device_operations {
+  retcode_t (*init)(void);            /**< initialize for device */
+  void (*fini)(void);                 /**< destructor for device  */
```
destructor of device

jserv commented on HAL.md:
````diff
@@ -0,0 +1,117 @@
+
+* Add device object after DEVICES_OBJ
+* Add new device build target
+```=
````
Don't set `=` here.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+## Modify hal Makefile
+Add new device into `hal/Makefile`:
+
+* Add device object after DEVICES_OBJ
```
Change `DEVICES_OBJ` to `DEVICE_OBJS`.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+static inline void unregister_emulator(void);
+...
+
+static struct device_operations emulator_ops = {.init = &emulator_init,
```
You shall indent.

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+};
+
+struct secure_store_operations {
+  retcode_t (*init)(void); /**< init secure storage */
```
In comments, always s/init/initialize/

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,135 @@
+struct device_operations {
+  retcode_t (*init)(void);            /**< initialize device */
+  void (*fini)(void);                 /**< destructor of device  */
+  retcode_t (*get_key)(uint8_t *);    /**< get device private key */
```
Don't mix `uint8_t` and `char`. In Arm/gcc, `char` defaults to be unsigned.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+
+static void emulator_release(void) { unregister_emulator(); }
+
+...
```
Avoid unexpected `...`

marktwtn commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+Here are some functions needed to use inside `impl.c`:
+* register_device : register device on startup
+* unregister_device : unregister device
+* DECLARE_DEVICE : this must be declare inside `impl.c`
```
`declared`.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,117 @@
+
+Implement new device operations, which are included from `hal/device.h`
+
+## Quick view inside device.h
```
What is "Quick view"? Can you use formal terms?

jserv commented on HAL.md:
````diff
@@ -0,0 +1,115 @@
+* Add device object after DEVICES_OBJS
+* Add new device build target
+```
+DEVICES_OBJS = wp7702.o emulator.o device.o mydevice.o
````
Should be `DEVICE_OBJS`

jserv commented on HAL.md:
```diff
@@ -0,0 +1,115 @@
+### Modify hal Makefile
+Add new device into `hal/Makefile`:
+
+* Add device object after DEVICES_OBJS
```
`DEVICE_OBJS` instead of `DEVICES_OBJS`

jserv commented on HAL.md:
````diff
@@ -0,0 +1,115 @@
+mydevice.o: impl.c
+        $(CC) $(CFLAGS) $(INCLUDES) -c $^ -o $@
+```
+`$(CC)`,`$(CFLAGS)` and `$(INCLUDES)` are provided by TA-endpoint. Don't define same flags inside Makefile.
````
You should propose the common rules rather than the confusing terms.

jserv commented on hal/Makefile:
```diff
@@ -0,0 +1,11 @@
+DEVICES_OBJS = device.o
+INCLUDES += -I/$(CURDIR)
```
Don't do that. Programmers should always include precise headers instead.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,115 @@
+    * write : write item to secure storage
+    * read : read item from secure storage
+    * delete : delete item inside secure storage
+### Functions to use inside `impl.c`
```
Delete the sections "Functions to use inside impl.c", "Modify hal Makefile", "Implement operators inside impl.c", etc. We should write down high-level descriptions first.

jserv replied:
Let's concentrate on qualified interface for further implementation.

jserv commented on HAL.md:
```diff
@@ -0,0 +1,115 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer supports such as uart, GPIO, secure storage and other low-level operators inside endpoint. The endpoint is a daemon based on Linux which offers monitor for the device and can send the message to tangle-accelerato ...
```
Should be "The endpoint is a daemon, offering monitor ..."

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,115 @@
+* third_party: Third party directory.
+* utils: Utility functions for TA-endpoint.
+
+## Implement new device
```
How to implement new device

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,115 @@
+
+## Implement new device
+
+### Create directory for new device
```
Create directory for the new device

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,115 @@
+## Implement new device
+
+### Create directory for new device
+Create directory for new device into `devices`.
```
Create directory for the new device under `devices`.

marktwtn commented on HAL.md:
````diff
@@ -0,0 +1,110 @@
+mkdir devices/mydevice
+```
+* Create `impl.c` and `Makefile` under new device directory.
+* Include `device.h` headers into new device header file.
````
Since you include `device.h` in the `impl.c` (as shown in the example), the description need to be modified from header file to source file.

marktwtn commented:
Make sure all the doxygen comment and source code related to `retcode_t` is added with the prefix `RET_`.

marktwtn commented on HAL.md:
```diff
@@ -0,0 +1,110 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer supports such as uart, GPIO, secure storage and other low-level operators inside endpoint. The endpoint is a daemon, offering monitor for the device and can send the message to tangle-accelerator. HAL layer lets  ...
```
`can send` -> `being able to send`
`which defined` -> `defined` or `which is defined`
`how endpoint work` -> `how endpoint works`

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,110 @@
+* devices: The real directory to implement devices.
+* hal: Hardware abstraction layer.
+* hooks: Git hook functions.
+* tests: Unit-tests.
```
I think the test under `/tests` will be more than unit-test in the future

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,110 @@
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer supports such as uart, GPIO, secure storage and other low-level operators inside endpoint. The endpoint is a daemon, offering monitor for the device and can send the message to tangle-accelerator. HAL layer lets  ...
+
+## Directory overview
+* connectivity: Connect to Tangle-accelerator functions.
```
It is the implementation of the communcation functions but not the function which connect to ta's functions

jserv commented on HAL.md:
```diff
@@ -0,0 +1,110 @@
+* hooks: Git hook functions.
+* tests: Test functions and unit-tests.
+* third_party: Third party directory.
+* utils: Utility functions for TA-endpoint.
```
Think of the merging process between TA and TA-endpoint. It is not necessary to make a list of directories at the moment. Instead, you can send further pull requests after this one is merged.

jserv replied:
HAL is a blocking issue for merge process though.

jserv commented:
I defer to @HowJMay .

Commit:
feat: Unify low-level operation with consistent HAL interface

Add hardware abstraction layer for create unify devices and
device operators. See `HAL.md` for how to implement device.

Closes #63

howjmay commented on HAL.md:
```diff
@@ -0,0 +1,100 @@
+# Hardware abstract layer(HAL) overview
+
+## Abstract
+Hardware abstract layer(HAL) defines a standard interface for hardware vendors to implement, which enables Tangle-Accelerator to be aware of lower-level driver implementations. HAL layer supports such as uart, GPIO, secure storage and other low-level operators inside endpoint. The endpoint is a daemon, offering monitor for the device and being able to send the message to tangle-accelerator. HAL l ...
```
for hardware vendors to implement [It is lack of a subject here], which enables

splasky replied:
Need a new subject here?

howjmay commented on HAL.md:
````diff
@@ -0,0 +1,100 @@
+mkdir devices/mydevice
+```
+* Create `impl.c` and `Makefile` under new device directory.
+* Include `device.h` headers into new device header or new device source.
````
header of header file ?

howjmay commented on HAL.md:
````diff
@@ -0,0 +1,100 @@
+mydevice.o: impl.c
+        $(CC) $(CFLAGS) $(INCLUDES) -c $^ -o $@
+```
+`$(CC)`,`$(CFLAGS)` and `$(INCLUDES)` are provided by TA-endpoint. `CC` sets the default compiler for project, `CFLAGS` are the default flags to be passed to default compiler during compiling, `INCLUDES` flag includes headers inside sub project and third-party library. You can also modify these flag inside your device Makefile.
````
$(CC)`,`$(CFLAGS)` and `$(INCLUDES)` are provided by TA-endpoint. `CC` sets the default compiler for the project. `CFLAGS` are the default flags that would be passed to default compiler during compiling time. `INCLUDES` flag includes headers inside sub project and third-party library. You can also modify these flags inside your device Makefile.

howjmay replied:
`sub diretory` and `device Mkefile` is somehow misleading or unclear. It is better to use ither words to replace them

jserv commented on utils/defined_error.h:
```diff
@@ -24,6 +24,14 @@ typedef enum {
   RET_HTTP_CERT,    /**< failed on x509 cert parse */
   RET_HTTP_CONNECT, /**< failed on HTTP initial connection */
   RET_HTTP_SSL,     /**< failed on setting ssl config */
+  RET_DEVICE_REG,   /**< register device error */
```
Don't use `RET_DEVICE_REG`. Instead, you can use `RET_DEVICE_INIT`. Similarly, `RET_DEVICE_FINI`.
@HowJMay, prefix `RET_` might be a bit misleading since we mention errors/warning here. How about changing it to `ERR_`?

splasky replied:
The prefix misleading problem has been discussed last week. This issue has been noted on todo list inside hackmd.

jserv replied:
`REG` is usually used for registers. Don't confuse.

jserv replied:
?

jserv replied:
If you are going to ask someone for reviewing, you must resolve all concerns during past discussions.
